src/ir_visitor/visitor.rs

```diff
@@ -217,27 +260,29 @@ where
         };
 
         // Traverse the commands
-        let mut cmds = std::mem::take(&mut comp.cmds);
-        match self.visit_cmds(&mut cmds, comp) {
+        let mut cmds = std::mem::take(&mut ctx.get_mut(idx).cmds);
```

Comment:
The way this is currently structured extracts the commands out from the component but gives you mutable access to the context. What I was thinking originally is that we instead completely "detach" the component from the context (by defining a `swap` method on `IndexStore`) and then providing mutable access to the component but immutable access to the context.

Reply:
To simplify the number of arguments then could we re-add the "idx()" function on components? BundleElim requires component idx information which currently would have to be passed in separately from everything else.

Reply:
Oh, I see. Yeah, we can try that but that was causing annoying build problems because we’d need a default idx

Reply:
One other option is to abstract all of this using a `VisitData` struct which is also more future proof than adding more arguments. Calyx visitor has the same problem where we've slowed made the visitor interface more and more complicated with more arguments

Reply:
Personally more of a fan of the second option! I'll work on that next.